### DISCUSSION TITLE
SAML single sign-on: RelayState is invalid

### DISCUSSION THREAD

**Discussion by superv01d** [community/community#110351] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

We're experiencing a "RelayState is invalid" error during our SAML SSO process, using a custom-developed IdP. 
Post request back to ACS contains 2 form parameters: SAMLResponse and RelayState. We've confirmed that the RelayState is unchanged.
What could be the problem? Any known issues or debugging tips would also be appreciated.

**Comment by crjamesl** [comment]
I just wanted to bump this thread.   I see the same issues coming from an IDP where it passes the relaystate in the SAML request the SSO page doesn't update its redirecturl as such and drops the signon'ed user into the default enterprise.

Has anyone found out how to fix this?

**Reply by YaiLung** [reply]
hello, the "Relay State is invalid" error in SAML SSO typically indicates that the RelayState parameter being sent back from the Identity Provider (IdP) to the Service Provider (SP) is not matching what the SP expects. 
Possible Causes & Solutions
Mismatch in RelayState Handling

The SP may expect a specific RelayState value or a signed/encoded version of it.
Ensure that the exact RelayState value sent by the SPin the authentication request is returned unchanged by the IdP.
RelayState Length Limitation

Some SAML implementations encode the RelayState as part of the request and expect it in the same format.
Ensure that URL encoding or Base64 encoding is correctly handled.
Compare the raw RelayState values sent and received to verify consistency.
Incorrect Handling of HTTP POST

If using an HTTP POST binding, make sure that the RelayState parameter is properly included in the response and not stripped or modified.
SP-side Validation Issues

**Reply by crjamesl** [reply]
I guess I should file a separate ticket.   I don't get any error.  It just isn't honored.   I see the relayState come through the SAML request and in the response it isn't returned or redirected on the github side.

I have tried linking to a /org/repo, full url path and url encoding both but nothing seems to redirect.   All saml requests just drop it back at the top of the enterprise level not processing the relayState.

**Reply by YaiLung** [reply]
the issue is likely on GitHub's side rather than with your IdP. Here are some possible reasons and steps to troubleshoot:

Possible Causes & Solutions
GitHub Does Not Support RelayState-Based Redirects

Some SPs (like GitHub Enterprise) may ignore RelayState or only honor it under certain conditions.
Check GitHub’s official SAML documentation to confirm if they support deep linking via RelayState.
GitHub’s Default Landing Behavior

GitHub may always redirect users to the enterprise homepage instead of honoring RelayState.
Test if it works when linking to an org-level URL (https://github.com/orgname), rather than a repo-specific one.
GitHub Enterprise vs. Cloud Differences

Even if the RelayState is present in your response, GitHub might expect it in a specific format.
Try encoding it differently (e.g., URL-encoding vs. Base64-encoding).

Debugging Steps
Confirm What GitHub Logs Show

Check the SAML audit logs in GitHub Enterprise to see if the RelayState is logged or mentioned.
Try a Minimal RelayState Value

Instead of a full URL, try a simple path like /org/repo.
If GitHub requires a full URL, ensure you’re using the correct base (https://github.com/org/repo).
Contact GitHub Support
